### ISSUE TITLE
Generic type param used as generic type constraint causes variance issues?

### ISSUE THREAD

**Issue by AnyhowStep** [microsoft/TypeScript#33561]
**TypeScript Version:**  3.5.1, 3.6.3, Nightly

**Search Terms:**

generic type parameter, generic type constraint, variance

**Code**

```ts
declare const x0 : <InT extends boolean|null>(arg : InT) => boolean;
//Expected: Error
//Actual  : Error
const x1 : <InT extends string>(arg : InT) => boolean = x0;

type Y<In, Out> = <InT extends In>(arg : InT) => Out;
declare const y0 : Y<boolean|null, boolean>;
//Expected: Error
//Actual  : OK
const y1 : Y<string, boolean> = y0;

```

**Expected behavior:**

`y1 = y0` assignment should fail

**Actual behavior:**

`y1 = y0` assignment passes

**Playground Link:** 

[3.5.1 Playground](http://www.typescriptlang.org/play/?ts=3.5.1#code/CYUwxgNghgTiAEYD2A7AzgF3gDwAzwC54AeASRQBV4RsMQVg14AjJJCEKFAHxQFcIEAHwAKWAHNC8chQCU8ALxCWbDlwDcAKAD02gKLYADuDrAiemDCQwd2gIJgMfKBHhSLVm8nRZsARikySmpaekZ4TBgASxRxUQkpGXklFXZOFEUcXC1NDABPY3gATSCAGngAeT4MZQUSGRC6BiZyeJhJIiTFZSqMLVBIWARvTHg8-CIS1jSuXgEIcum1FCEtXQNjRxAzeA9rWwcnFzciCoBpTRGsPIDJ4kiY8UXVdNqx7M0gA)

[3.6.3 Playground](http://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEYD2A7AzgF3gDwAzwC54AeASRQBV4RsMQVg14AjJJCEKFAHxQFcIEAHwAKWAHNC8chQCU8ALxCWbDlwDcAKAD02gKLYADuDrAiemDCQwd2gIJgMfKBHhSLVm8nRZsARikySmpaekZ4TBgASxRxUQkpGXklFXZOFEUcXC1NDABPY3gATSCAGngAeT4MZQUSGRC6BiZyeJhJIiTFZSqMLVBIWARvTHg8-CIS1jSuXgEIcum1FCEtXQNjRxAzeA9rWwcnFzciCoBpTRGsPIDJ4kiY8UXVdNqx7M0gA)

[Nightly Playground](http://www.typescriptlang.org/play/?ts=Nightly#code/CYUwxgNghgTiAEYD2A7AzgF3gDwAzwC54AeASRQBV4RsMQVg14AjJJCEKFAHxQFcIEAHwAKWAHNC8chQCU8ALxCWbDlwDcAKAD02gKLYADuDrAiemDCQwd2gIJgMfKBHhSLVm8nRZsARikySmpaekZ4TBgASxRxUQkpGXklFXZOFEUcXC1NDABPY3gATSCAGngAeT4MZQUSGRC6BiZyeJhJIiTFZSqMLVBIWARvTHg8-CIS1jSuXgEIcum1FCEtXQNjRxAzeA9rWwcnFzciCoBpTRGsPIDJ4kiY8UXVdNqx7M0gA)

**Related Issues:**

**Comment by AnyhowStep** [comment]
Seems like the type inference is OK, though,

```ts
/**
 * type test0 = [boolean | null, boolean]
 */
type test0 = typeof y0 extends Y<infer In, infer Out> ?
    [In, Out] :
    never;

/**
 * type test1 = [string, boolean]
 */
type test1 = typeof y1 extends Y<infer In, infer Out> ?
    [In, Out] :
    never;
```
[3.5.1 Playground](http://www.typescriptlang.org/play/?ts=3.5.1#code/CYUwxgNghgTiAEYD2A7AzgF3gDwAzwC54AeASRQBV4RsMQVg14AjJJCEKFAHxQFcIEAHwAKWAHNC8chQCU8ALxCWbDlwDcAKAD02gKLYADuDrAiemDCQwd2gIJgMfKBHhSLVm8nRZsARikySmpaekZ4TBgASxRxUQkpGXklFXZOFEUcXC1NDABPY3gATSCAGngAeT4MZQUSGRC6BiZyeJhJIiTFZSqMLVBIWARvTHg8-CIS1jSuXgEIcum1FCEtXQNjRxAzeA9rWwcnFzciCoBpTRGsPIDJ4kiY8UXVdNqx7M0dACovzXgv+D5Qp0TD4OoAbSW6Xg3Hg-EEzxmKAAun8vtpcgUECCMGDAVikAAzd6NMJMEoxQkgGDSFDlSnUyrVZQAfj+bng4PI5V6yMI7LcKBAADdqTltD80fjgSBMAEIQ9Yojlqj-higdjZRh5dKQESxgEaE1whSUFSadz4Ayab1WQLOZbefyOXCRWKgA)

**Comment by jack-williams** [comment]
Duplicate of #31006 and fixed by #31029.

**Comment by AnyhowStep** [comment]
It's not in the nightly builds?
